xunit.runner.visualstudio.desktop/VsTestRunner.cs

```diff
@@ -445,12 +459,83 @@ void RunTests(IRunContext runContext, IFrameworkHandle frameworkHandle, LoggerHe
+                                if (discoveryOptions.GetDiagnosticMessagesOrDefault())
+                                    foreach (var testCase in visitor.TestCases)
+                                        logger.Log(testCase, "Discovered [execution] test case '{0}' (ID = '{1}')",
+                                            testCase.DisplayName, testCase.UniqueID);
+
+                                assemblyDiscoveredInfo = new AssemblyDiscoveredInfo
+                                {
+                                    AssemblyFileName = source,
+                                    DiscoveredTestCases = visitor.TestCases.Select(testCase => new DiscoveredTestCase(source, discoverer, testCase, logger, testPlatformContext)).ToList()
+                                };
+                            },
+                            assemblyFileName,
+                            shadowCopy,
+                            configuration
+                        );
+
+                        if (assemblyDiscoveredInfo.DiscoveredTestCases == null ||
+                            !assemblyDiscoveredInfo.DiscoveredTestCases.Any())
+                        {
+                            if (configuration.DiagnosticMessagesOrDefault)
+                            {
+                                logger.LogWarning("Skipping '{0}' since no tests were found during discovery [execution].", assemblyDiscoveredInfo.AssemblyFileName);
+                            }
+                            return;
+                        }
+
+                        // Filter tests
+                        var traitNames = new HashSet<string>(assemblyDiscoveredInfo.DiscoveredTestCases.SelectMany(testCase => testCase.TraitNames));
+
+                        // Apply any filtering
+                        var filter = new TestCaseFilter(runContext, logger, assemblyDiscoveredInfo.AssemblyFileName, traitNames);
+                        var filteredTestCases = assemblyDiscoveredInfo.DiscoveredTestCases.Where(dtc => filter.MatchTestCase(dtc.VSTestCase)).ToList();
+
+                        // Force unique names if there is more than 1 testcase with the same name
+                        foreach (var groupWithDuplicateNames in filteredTestCases.GroupBy(dtc => dtc.Name).Where(group => group.Count() > 1))
+                        {
+                            foreach (var discoveredTestCaseWithDuplicateName in groupWithDuplicateNames)
+                                discoveredTestCaseWithDuplicateName.ForceUniqueName();
+                        }
 
+                        foreach (var filteredTestCase in filteredTestCases)
+                        {
+                            // PERF: the filteredTestCase.TestCase object is proxy to ITestCase in the discovery appdomain.
+                            // We're using the UniqueId only (cached in this appdomain).
+                            testCasesMap.Add(filteredTestCase.UniqueId, filteredTestCase.VSTestCase);
```

Comment:
We pre-create a map of `UniqueID, VSTestCase` to allow faster lookups in `VSExecutionSink`.